File: server/webapp/WEB-INF/rails.new/app/assets/new_javascripts/models/pipeline_configs/tracking_tool.js
/*
 * Copyright 2016 ThoughtWorks, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

define(['mithril', 'lodash', 'string-plus', 'models/model_mixins', 'models/validatable_mixin'], function (m, _, s, Mixins, Validatable) {

  var UrlPatternValidator = function () {
    this.validate = function (entity) {
      if (!s.include(entity.urlPattern(), '${ID}')) {
        entity.errors().add('urlPattern', Validatable.ErrorMessages.mustContainString("urlPattern", '${ID}'));
      }
    };
  };

  var TrackingTool = function (type) {
    this.constructor.modelType = 'trackingTool';
    Mixins.HasUUID.call(this);

    this.type = m.prop(type);

    var self = this;

    this.toJSON = function () {
      return {
        type:       self.type(),
        attributes: self._attributesToJSON()
      };
    };

    this._attributesToJSON = function () {
      throw new Error("Subclass responsibility!");
    };
  };

  TrackingTool.Generic = function (data) {
    TrackingTool.call(this, "generic");
    Validatable.call(this, data);
    this.urlPattern = m.prop(s.defaultToIfBlank(data.urlPattern, ''));
    this.regex      = m.prop(s.defaultToIfBlank(data.regex, ''));

    this.validatePresenceOf('urlPattern');
    this.validateUrlPattern('urlPattern');
    this.validateWith('urlPattern', UrlPatternValidator);
    this.validatePresenceOf('regex');

    this._attributesToJSON = function () {
      return {
        urlPattern: this.urlPattern(),
        regex:      this.regex()
      };
    };
  };

  TrackingTool.Generic.fromJSON = function (data) {
    return new TrackingTool.Generic({
      urlPattern: data.url_pattern,
      regex:      data.regex
    });
  };

  TrackingTool.Mingle = function (data) {
    TrackingTool.call(this, "mingle");
    Validatable.call(this, data);
    this.baseUrl               = m.prop(s.defaultToIfBlank(data.baseUrl, ''));
    this.projectIdentifier     = m.prop(s.defaultToIfBlank(data.projectIdentifier, ''));
    this.mqlGroupingConditions = m.prop(s.defaultToIfBlank(data.mqlGroupingConditions, ''));

    this.validatePresenceOf('baseUrl');
    this.validateUrlPattern('baseUrl');
    this.validatePresenceOf('projectIdentifier');
    this.validatePresenceOf('mqlGroupingConditions');

    this._attributesToJSON = function () {
      return {
        baseUrl:               this.baseUrl(),
        projectIdentifier:     this.projectIdentifier(),
        mqlGroupingConditions: this.mqlGroupingConditions()
      };
    };

  };

  TrackingTool.Mingle.fromJSON = function (data) {
    return new TrackingTool.Mingle({
      baseUrl:               data.base_url,
      projectIdentifier:     data.project_identifier,
      mqlGroupingConditions: data.mql_grouping_conditions
    });
  };

  TrackingTool.Types = {
    generic: {type: TrackingTool.Generic, description: "Generic"},
    mingle:  {type: TrackingTool.Mingle, description: "Mingle"}
  };

  TrackingTool.create = function (type) {
    return new TrackingTool.Types[type].type({});
  };

  TrackingTool.fromJSON = function (data) {
    if (!_.isEmpty(data)) {
      return TrackingTool.Types[data.type].type.fromJSON(data.attributes || {});
    }
  };
  return TrackingTool;
});

File: config/config-api/src/main/java/com/thoughtworks/go/config/TrackingTool.java
/*
 * Copyright 2015 ThoughtWorks, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.thoughtworks.go.config;

import java.util.Map;

import com.thoughtworks.go.domain.CommentRenderer;
import com.thoughtworks.go.domain.DefaultCommentRenderer;
import com.thoughtworks.go.domain.ConfigErrors;
import org.apache.commons.lang.StringUtils;

@ConfigTag("trackingtool")
public class TrackingTool implements ParamsAttributeAware, Validatable, CommentRenderer {
    @ConfigAttribute(value = "link", optional = false)
    private String link = "";
    @ConfigAttribute(value = "regex", optional = false)
    private String regex = "";

    public static final String LINK = "link";
    public static final String REGEX = "regex";
    private ConfigErrors configErrors = new ConfigErrors();


    public TrackingTool() {
    }

    public TrackingTool(String link, String regex) {
        this.link = link;
        this.regex = regex;
    }

    public String getLink() {
        return link;
    }

    public void setLink(String link) {
        this.link = link;
    }

    public String getRegex() {
        return regex;
    }

    public void setRegex(String regex) {
        this.regex = regex;
    }

    public String render(String text) {
        return new DefaultCommentRenderer(link, regex).render(text);
    }

    public void setConfigAttributes(Object attributes) {
        if (attributes == null) return;
        Map attributeMap = (Map) attributes;
        if (attributeMap.containsKey(LINK)) {
            link = (String) attributeMap.get(LINK);
        }
        if (attributeMap.containsKey(REGEX)) {
            regex = (String) attributeMap.get(REGEX);
        }
    }

    public boolean validateTree(ValidationContext validationContext) {
        validate(validationContext);
        return errors().isEmpty();
    }

    public void validate(ValidationContext validationContext) {
        if (StringUtils.isEmpty(link)) {
            configErrors.add(LINK, "Link should be populated");
        }
        if (StringUtils.isEmpty(regex)) {
            configErrors.add(REGEX, "Regex should be populated");
        }
        if (!link.contains("${ID}")) {
            configErrors.add(LINK, "Link must be a URL containing '${ID}'. Go will replace the string '${ID}' with the first matched group from the regex at run-time.");
        }
    }

    public ConfigErrors errors() {
        return configErrors;
    }

    public void addError(String fieldName, String message) {
        configErrors.add(fieldName, message);
    }

    public static TrackingTool createTrackingTool(Map attributes) {
        TrackingTool trackingTool = new TrackingTool();
        trackingTool.setConfigAttributes(attributes);
        return trackingTool;
    }

    public boolean isDefined() {
        return !link.isEmpty();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        TrackingTool that = (TrackingTool) o;
        if (link != null ? !link.equals(that.link) : that.link != null) {
            return false;
        }
        return !(regex != null ? !regex.equals(that.regex) : that.regex != null);

    }

    @Override
    public int hashCode() {
        int result = link != null ? link.hashCode() : 0;
        result = 31 * result + (regex != null ? regex.hashCode() : 0);
        return result;
    }
}
